arcgis-ios-sdk-samples/Layers/Browse OGC API feature service/BrowseOGCAPIServicesViewController.swift

```diff
@@ -0,0 +1,224 @@
+        // panning and zooming won't request features automatically.
+        table.featureRequestMode = .manualCache
+        
+        lastQuery = table.populateFromService(
+            with: queryParameters,
+            clearCache: false,
+            outfields: nil
+        ) { [weak self, table] _, error in
+            guard let self = self else { return }
+            if let error = error,
+               // Do not display error if user simply cancelled the request.
+               (error as NSError).code != NSUserCancelledError {
+                self.presentAlert(error: error)
+            } else if let extent = info.extent {
+                // Zoom to the extent of the selected collection.
+                let featureLayer = AGSFeatureLayer(featureTable: table)
+                featureLayer.renderer = self.makeRenderer(for: table.geometryType)
+                self.map.operationalLayers.add(featureLayer)
+                self.mapView.setViewpointGeometry(extent, padding: 50, completion: nil)
+            }
+        }
+    }
+    
+    /// Create an appropriate renderer for a type of geometry.
+    func makeRenderer(for geometryType: AGSGeometryType) -> AGSSimpleRenderer? {
+        let renderer: AGSSimpleRenderer?
+        switch geometryType {
+        case .point, .multipoint:
+            renderer = AGSSimpleRenderer(symbol: AGSSimpleMarkerSymbol(style: .circle, color: .blue, size: 5))
+        case .polyline:
+            renderer = AGSSimpleRenderer(symbol: AGSSimpleLineSymbol(style: .solid, color: .blue, width: 1))
+        case .polygon, .envelope:
+            renderer = AGSSimpleRenderer(symbol: AGSSimpleFillSymbol(style: .solid, color: .blue, outline: nil))
+        default:
+            renderer = nil
+        }
+        return renderer
+    }
+    
+    // MARK: Action
+    
+    func askUserForServiceURL(completion: @escaping (Result<URL, Error>) -> Void) {
+        // Create an object to observe changes from the text fields.
+        var textFieldObserver: NSObjectProtocol!
+        let alertController = UIAlertController(
+            title: "Load OGC API feature service",
+            message: "Please provide a URL to an OGC API feature service.",
+            preferredStyle: .alert
+        )
+        // Remove observer on cancel.
+        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel) { _ in
+            NotificationCenter.default.removeObserver(textFieldObserver!)
+            completion(.failure(InputError.userCancelled))
+        }
+        // Validate and create URL from user input and remove observer.
+        let loadAction = UIAlertAction(title: "Load", style: .default) { _ in
+            NotificationCenter.default.removeObserver(textFieldObserver!)
+            guard let text = alertController.textFields![0].text?.trimmingCharacters(in: .whitespacesAndNewlines),
+                  let serviceURL = URL(string: text) else {
+                completion(.failure(InputError.invalidURL))
```

Comment:
When an invalid URL is inputted, I think it'd be nice to prompt the user with another text field so they can try another URL. This would be useful for typos so the user wouldn't have to exit the sample to try a new URL.

Reply:
Please take a look at the new changes and share your thoughts. 🙂 

The code is a bit ugly because it's a recursive closure in this case. Also, do you think having a 2-step cancel makes sense?

Reply:
Per Phil I'll remove the retry logic. I'll re-request your review when it's updated!